### ISSUE TITLE
SQL0444N Reason Code 5

### ISSUE THREAD

**Issue by mknelsen** [tessus/db2-hash-routines#4]
Hi, I've tried to install these routines on a db2 v 10.5 server, and I'm getting the following error

SQL0444N  Routine "APR_SHA256" (specific name "SP_APR_SHA256") is implemented
with code in library or path ".../sqllib/function/unfenced/hash", function
"aprsha256" which cannot be accessed.  Reason code: "5".  SQLSTATE=42724

Just wondering if you have seen this error or know how I can troubleshoot this?

**Comment by tessus** [comment]
Reason code 5 means the following (you get that info by running `db2 "? SQL0444N"` on the command line):

```
5

         There is insufficient memory to load the library containing the
         function or one or more symbols could not be resolved. This
         reason code indicates one of the following situations:


         1. One or more symbols might not have been resolved. The
            routine library might be dependent on a shared library that
            cannot be located (using the concatenation of directories
            specified in the LIBPATH environment variable in UNIX-based
            systems, the PATH environment variable in INTEL systems).
         2. The routine has a 64-bit library which is not supported
            within a 32-bit DB2 instance, or, the routine has a 32-bit
            library or DLL within a 64-bit DB2 instance that is
            incompatible with the routine definition.
         3. There was insufficient memory to load the library containing
            the function.
```
Possible solutions include:

```
5

         Possible solutions are ordered to match the causes provided in
         the message explanation:


         1. There can be multiple levels of this indirection involved,
            leading to this reason code. For example, routine body X can
            be found, and it requires shared library Y which is also
            found, however Y needs another shared library Z that cannot
            be located. The routine library should be relinked or the
            user should verify that all dependent libraries are
            available in the LIBPATH (UNIX) or PATH (INTEL). Use the
            db2set command to update the LIBPATH to include the path
            where the library is located. For example: db2set
            db2libpath=/db2test/db2inst1/sqll
         2. Use a supported library.
         3. If a lack of memory is determined then the system
            configuration may need to be changed to make more memory
            available to DB2.
```

Have a look at the `db2diag.log` in the `db2dump` directory. You will most likely find additional info in there.
Also, what happens, if you run the `test_hash` program?

**Comment by tessus** [comment]
Add the `apr` and `apr-util` library paths to the `/etc/ld.so.conf` and run `ldconfig`.

You shouldn't have to do this, since the linker should have been given this info by running the `makertn` script.

**Comment by mknelsen** [comment]
Thanks for your reply.  I'm a little over my head here, I'm more of a db2 guy then a UNIX guy.  

I ran 
gcc test_hash.c 
And got the following.

hash.h:27:17 fatal error: apr.h: No such file or directory
#include "apr.h"

I'm going to reach out to some UNIX people I know to see if they can help me as well.

Thanks again

Mike

**Comment by tessus** [comment]
It seems you are missing the apr-devel and apr-util-devel packages. APR and APR-util are a requirement for the routines.
Page 1 of the [documentation](http://tessus.github.io/documentation/db2-hash-routines.pdf) also describes the build process.

If you can't fix the problem yourself, please send me an email.

**Comment by mknelsen** [comment]
Ya, I forgot to run the compiler first.

Ran the maketest script and the test_hash executable and they work just fine.  It looks like it's something with db2.  

Here's the info from db2 logs 

 
2016-01-04-10.06.26.123989-360 E67385805A916        LEVEL: Error (OS)
PID     : 12058690             TID : 22055          PROC : db2sysc 0
INSTANCE: db2inst2             NODE : 000           DB   : PHOENIX
APPHDL  : 0-2677               APPID: *LOCAL.db2inst2.160104131643
AUTHID  : DB2INST2             HOSTNAME: ab22-dwhts001
EDUID   : 22055                EDUNAME: db2agent (PHOENIX) 0
FUNCTION: DB2 UDB, oper system services, sqloAIXLoadModuleTryShr, probe:130
CALLED  : OS, -, dlopen
OSERR   : ENOEXEC (8) "Exec format error"
MESSAGE : Attempt to load specified library failed.
DATA #1 : Library name or path, 44 bytes
/home/db2inst2/sqllib/function/unfenced/hash
DATA #2 : shared library load flags, PD_TYPE_LOAD_FLAGS, 4 bytes
2
DATA #3 : String, 151 bytes
        0509-022 Cannot load module /home/db2inst2/sqllib/function/unfenced/hash.
        0509-026 System error: Cannot run a file that does not have a valid format.
 
2016-01-04-10.06.26.124923-360 E67386722A1093       LEVEL: Error (OS)
PID     : 12058690             TID : 22055          PROC : db2sysc 0
INSTANCE: db2inst2             NODE : 000           DB   : PHOENIX
APPHDL  : 0-2677               APPID: *LOCAL.db2inst2.160104131643
AUTHID  : DB2INST2             HOSTNAME: ab22-dwhts001
EDUID   : 22055                EDUNAME: db2agent (PHOENIX) 0
FUNCTION: DB2 UDB, oper system services, sqloAIXLoadModuleTryShr, probe:140
CALLED  : OS, -, dlopen
OSERR   : ENOEXEC (8) "Exec format error"
MESSAGE : Attempt to load specified library augmented with object name failed.
DATA #1 : Library name or path, 54 bytes
/home/db2inst2/sqllib/function/unfenced/hash(shr_64.o)
DATA #2 : shared library load flags, PD_TYPE_LOAD_FLAGS, 4 bytes
262146
DATA #3 : String, 286 bytes
        0509-022 Cannot load module /home/db2inst2/sqllib/function/unfenced/hash(shr_64.o).
        0509-153   File /home/db2inst2/sqllib/function/unfenced/hash is not an archive or
                   the file could not be read properly.
        0509-026 System error: Cannot run a file that does not have a valid format.

**Comment by tessus** [comment]
Yes, it seems that something went wrong with creating the shared library. Unfortunately I don't have access to an AIX box to investigate this further.
Are you using gcc or xlc? Did you compile APR and APR-util (plus their dev packages) yourself or did you install rpms?

I know that the UDFs and SPs run on AIX, since I have tested them on AIX.
Since I don't have access, please show the following errors to your AIX guys and they should be able to fix it.

```
0509-022 Cannot load module /home/db2inst2/sqllib/function/unfenced/hash(shr_64.o).
0509-153 File /home/db2inst2/sqllib/function/unfenced/hash is not an archive or the file could not be read properly. 
0509-026 System error: Cannot run a file that does not have a valid format.
```

**Comment by mknelsen** [comment]
Do you know if this has been used on db2 v 10.5?  I'm wondering if the newer version of db2 could be the issue?

**Comment by tessus** [comment]
The DB2 API hasn't changed, so they will work on all DB2 LUW releases.

As I mentioned before, I don't have access to an AIX machine. If your AIX guys can't help you, I can offer you to have a look. That is, if you can give me ssh access to a test instance.

**Comment by mknelsen** [comment]
I think we tracked down the issue.  It looks like when it compiles, it is making a 32 bit file, and we are running a 64 bit database.  I'm checking now to see if the Apache stuff has a 64 bit library or not, since when I force the program to compile as 64, it's complaining about a libapr-1.so being 32 bit

Mike

**Comment by tessus** [comment]
Hey Mike. 

Let me guess, You installed the 32bit APR* rpms and did not compile them yourself.
Easy solution. Download the apr and apr-util sources from http://apr.apache.org and compile them with the `configure` option `--prefix=/usr/local/apr`.
Then, point to your compiled version in the makertn script. Done.

**Comment by tessus** [comment]
@mknelsen any update? were you able to fix it?

**Comment by mknelsen** [comment]
Still working on it.  I've been dragged into other issues that are higher priority then this one right now.

**Comment by mknelsen** [comment]
I was finally able to get this to work, but I actually had to move it to a different machine to compile it.  When I compile the test hash it test hash route, it comes out to a size of 500000, but the hash only comes to 21000 on my system, where as on the other system, it works out closer in size to the test hash.

Between the 2 systems, there are a few differences, I'm running aix 7.1 with db2 10.5 and he's running aix 6.3 and db2 9.7, so it's hard to say what's wrong on my size, but at least it's working.  Thanks for you help.

Mike

**Comment by tessus** [comment]
I'm glad you were able to solve it.

I compiled the routine on AIX 7 with DB2 10.5 about 2 years ago, but I remember that AIX is quite picky about mixed libraries and/or differently linked shared objects.
I used `gcc` and `xlc_r`. But what I usually did was to stick to one compiler for all components (APR, APR-util, routine).

Your size inconsistency looks like something's off with linking (static vs dynamic). Very strange.
But it is supposed to be rather small, since it should create a dynamic shared object.
I really would need access to a current AIX box to investigate what went wrong on your side.

Maybe I should add some debug output to the `makertn` script to make problem determination easier. I'll put that in my TODO list.
